tony-history-server/test/utils/TestHdfsUtils.java

```diff
@@ -76,38 +77,106 @@ public void testContentOfHdfsFile_throwsException() throws IOException {
+    Path histFolder = new Path("./test/resources/emptyHistFolder");
+
+    assertEquals(new ArrayList<Path>(), HdfsUtils.getJobFolders(fs, histFolder, "job*"));
+  }
+
+  @Test
+  public void testGetJobFolders_typicalHistoryFolder() {
+    Path histFolder = new Path("./test/resources/typicalHistFolder");
+    String regex = "^job.*";
     List<Path> expectedRes = new ArrayList<>();
 
     for (int i = 1; i < 6; ++i) {
       StringBuilder sb = new StringBuilder();
       sb.append("file:").append(System.getProperty("user.dir"));
       sb.append("/test/resources/typicalHistFolder/job").append(i);
-      sb.append("/metadata.json");
       expectedRes.add(new Path(sb.toString()));
     }
-    List<Path> actualRes = HdfsUtils.getMetadataFilePaths(fs, histFolder);
+    List<Path> actualRes = HdfsUtils.getJobFolders(fs, histFolder, regex);
     Collections.sort(actualRes);
 
     assertEquals(expectedRes, actualRes);
     assertEquals(expectedRes.size(), actualRes.size());
   }
 
   @Test
-  public void testGetMetadataFilePaths_throwsException() throws IOException {
-    String histFolder = "./test/resources/typicalHistFolder";
+  public void testGetJobFolders_nestedHistFolder() {
+    Path histFolder = new Path("./test/resources/nestedHistFolder");
+    String regex = "^job.*";
+    List<Path> expectedRes = new ArrayList<>();
+    StringBuilder sb = new StringBuilder();
+    sb.append("file:").append(System.getProperty("user.dir"));
+    sb.append("/test/resources/nestedHistFolder/");
+    String baseDir = sb.toString();
+
+    expectedRes.add(new Path(baseDir, "2018/01/02/job0"));
+    expectedRes.add(new Path(baseDir,"2018/01/01/job1"));
+    expectedRes.add(new Path(baseDir,"2018/12/31/job2"));
+    expectedRes.add(new Path(baseDir,"2017/07/job3"));
+    expectedRes.add(new Path(baseDir,"2017/07/job4"));
+    expectedRes.add(new Path(baseDir,"2017/07/job5"));
+
+    List<Path> actualRes = HdfsUtils.getJobFolders(fs, histFolder, regex);
+    Collections.sort(actualRes, (o1, o2) -> {
+      String job1 = HdfsUtils.getJobId(o1.toString());
+      String job2 = HdfsUtils.getJobId(o2.toString());
+      return job1.charAt(job1.length()-1) - job2.charAt(job2.length()-1);
+    });
+
+    assertEquals(expectedRes, actualRes);
+    assertEquals(expectedRes.size(), actualRes.size());
+  }
+
+  @Test
+  public void testGetJobFolders_throwsException() throws IOException {
+    Path histFolder = new Path("./test/resources/typicalHistFolder");
     FileSystem mockFs = mock(FileSystem.class);
+    String regex = "job*";
 
-    when(mockFs.listStatus(new Path(histFolder))).thenThrow(new IOException("IO Excpt"));
-    List<Path> actualRes = HdfsUtils.getMetadataFilePaths(mockFs, histFolder);
+    when(mockFs.listStatus(histFolder)).thenThrow(new IOException("IO Excpt"));
+    List<Path> actualRes = HdfsUtils.getJobFolders(mockFs, histFolder, regex);
 
     assertEquals(new ArrayList<Path>(), actualRes);
     assertEquals(0, actualRes.size());
```

Comment:
Having both the line 181 and 182 checks seems redundant. I think the size 0 check on line 182 is sufficient. Also, it's a little weird to be asserting something about an implementation deal (that the returned List is specifically an `ArrayList`) when the getJobFolders() method API just says it returns a `List`.

Reply:
Ok that's fair. Fixed: https://github.com/linkedin/TonY/pull/81/commits/108fe8650f76b5d9d5ae71188f3b95d170b54a93.